### ISSUE TITLE
[css-text] Ability to control hyphenation of proper nouns/capitalised words

### ISSUE THREAD

**Issue by buttercookie42** [w3c/csswg-drafts#5157]
As of #3927, the css-text spec has gained [the guidance](https://drafts.csswg.org/css-text-3/#valdef-hyphens-auto) that
> In some languages (such as English but not German), it may [be] appropriate to avoid having hyphenation opportunities in mixed case words, as those may indicate proper nouns

and Gecko has since implemented something like that.
That's all very nice, but now I've run precisely into [the concerns voiced here](https://github.com/w3c/csswg-drafts/issues/3927#issuecomment-492121209) [and here](https://github.com/w3c/csswg-drafts/issues/3927#issuecomment-492208808), namely that the default behaviour of the browser doesn't fit my own requirements [1], but I have currently no way of overriding it.

So as per
> Note that systems such as TeX (the `\uchyph` parameter) and InDesign (the "Hyphenate Capitalized Words" option in paragraph formatting) do expose this question to authors, recognizing that there is not a simple "correct" behavior that the application can universally use.

I'd like to have the same ability in CSS as well.

[1] I have a very narrow column of text, so

1. I am prepared to accept hyphenation even of proper nouns as preferable to the behaviour of `overflow-wrap: break-word`
2. For my content, automatic hyphenation seems to work well enough, so enabling automatic hyphenation and fixing up some possible instances of over-eager hyphenation with a Word Joiner is less work than having to sprinkle around soft hyphens everywhere.
3. That column is mostly written in Title Case, so any implementation basing its decisions on capitalisation will misdetect all those words as proper nouns and disable hyphenation without me actually wanting to.

**Comment by frivoal** [comment]
Level 3 doesn't force the user agent to behave either way, merely notes that capitalization is one of the thing a UA *may* consider when deciding to hyphenate or not.

Given that TeX and InDesign offer controls to authors over this, I think it is reasonable to consider the possibility for level 4.

I think that the default will necessarily have to be an "auto" value, to let browsers use the best heuristics they know of, but what should the other values be?

**Comment by buttercookie42** [comment]
Something along the lines of `auto | always | never` possibly?
Certainly `always` (or some equivalent naming) to always allow automatic hyphenation of words containing capital letters, since that is what I was asking for, but a case could also be made for explicit control in the other direction, i.e. never automatically hyphenate words containing capital letters:

1. For some languages (like German, where *all* nouns are capitalised) the default behaviour of `auto` would probably rather correspond to `always`, so for manual control you'd also need an opposite option.
2. @jfkthame [mentioned](https://github.com/w3c/csswg-drafts/issues/3927#issuecomment-492121209) that one possible future refinement of Gecko's default behaviour (corresponding to a future `auto` value) might be to vary the decision based on line length, i.e. be less strict about not hyphenating capitalised words for short lines. In that case full manual control again would require both `always` and `never` values.

**Comment by jfkthame** [comment]
Something like `hyphenate-capitalized: auto | always | never` sounds good to me. Or maybe shorten it to `hyphenate-caps`, noting that the form `caps` is already used in `font-variant-caps`.

**Comment by frivoal** [comment]
`always` is probably the wrong term. If the UA wouldn't allow hyphenation in the word for other reasons (too short, not in the dictionary…), this value won't create hyphenation opportunity.
Maybe `auto | allow | forbid`?

**Comment by jfkthame** [comment]
Sure, I'd be fine with that too (although I didn't feel that `always` implies hyphenation will suddenly happen to capitalized words that couldn't normally be hyphenated even without caps).

**Comment by jfkthame** [comment]
Maybe even simpler, `auto | yes | no` ?

**Comment by buttercookie42** [comment]
Good point, yes.

And for completeness to make it clear and put it up for discussion: My understanding of the idea is that this should only affect the behaviour of `hyphens: auto`. I.e. `forbid` + `hyphens: auto` (or `manual` for that matter, not that that combination would make much sense then) would still consider hyphenation opportunities introduced by explicit (soft-)hyphens.

**Comment by jfkthame** [comment]
Agreed, this is only about whether `hyphens: auto` gets to attempt hyphenation. Explicit hyphens or &shy; should not be affected. If the author put them there, presumably they're intended to be used.

**Comment by mnater** [comment]
> this is only about whether hyphens: auto gets to attempt hyphenation. Explicit hyphens or &shy; should not be affected. If the author put them there, presumably they're intended to be used.

I beg to differ.
a) There are JavaScript tools that use soft hyphens everywhere, so they are not necessarily put there with intention.
b) The other hyphenation control properties seem to apply to both types of hyphenation (automatic or with soft hyphens, see https://github.com/w3c/csswg-drafts/issues/5090).

In my opinion the new property should work in both cases...

**Comment by buttercookie42** [comment]
Then for completeness you'd have to have some sort of `manual` value (never automatically hyphenate capitalised words, but still respect soft-hyphens), too, wouldn't you?

**Comment by jfkthame** [comment]
> > this is only about whether hyphens: auto gets to attempt hyphenation. Explicit hyphens or ­ should not be affected. If the author put them there, presumably they're intended to be used.
> 
> I beg to differ.
> a) There are JavaScript tools that use soft hyphens everywhere, so they are not necessarily put there with intention.

JavaScript tools that insert soft-hyphens have access to the text, and can make their own decisions whether to insert them in capitalized words or not. That (IMO) is out of scope for what the CSS hyphenation properties are seeking to control.

> b) The other hyphenation control properties seem to apply to both types of hyphenation (automatic or with soft hyphens, see #5090).

For the (proposed/draft) properties that provide control over where hyphenation is used within the overall layout, yes. For properties that fine-tune where the hyphenation algorithm may insert potential breaks, that's less clear; I think such properties should not affect manual soft hyphens. (See https://github.com/w3c/csswg-drafts/issues/5090#issuecomment-639407007.)

**Comment by HeikkiYlipaavalniemi** [comment]
it would be really good to be able to control also for words which are starting a sentence with a capital letter.

Both Finnish and Swedish languages can have really long words and in a mobile layout they can easily go over the screen width if they are starting a sentence. Either it will just wrap to the next line without hyphenation or break the layout.
